### ISSUE TITLE
Regularization not working in most recent git version?

### ISSUE THREAD

**Issue by fhieber** [VowpalWabbit/vowpal_wabbit#307]
Hi,

I am quite new to vw, so excuse me if this is a silly question, but it seems to me that regularization (at least l1) is not working in the current version of the code. Even though STDERR reports usage of l1 ('using l1 regularization = 100') result of training is exactly (!) the same as without l1 regularization -- which seems unlikely for l1_lambda=100.

> /toolbox/vowpal_wabbit/vw --sgd  --noconstant -c -f out_l1 --l1 100 --passes 1  --readable_model out_l1.readable -k < x0
> using l1 regularization = 100
> Num weight bits = 18
> learning rate = 10
> initial_t = 1
> power_t = 0.5
> final_regressor = out_l1
> creating cache_file = .cache
> Reading datafile = 
> num sources = 1
> average    since         example     example  current  current  current
> loss       last          counter      weight    label  predict features
> 4.000000   4.000000            1         1.0   2.0000   0.0000      104
> ...
> 1.118318   0.962480       524288    524288.0   2.0000   2.0007        5
> 
> finished run
> number of examples per pass = 807449
> passes used = 1
> weighted example sum = 807449
> weighted label sum = 1.6647e+06
> average loss = 1.01294
> best constant = 2.06168
> total feature number = 22483110

Here's the run without l1:

> vw --sgd  --noconstant -c -f out --passes 1  --readable_model out.readable -k < x0
> Num weight bits = 18
> learning rate = 10
> initial_t = 1
> power_t = 0.5
> final_regressor = out
> creating cache_file = .cache
> Reading datafile = 
> num sources = 1
> average    since         example     example  current  current  current
> loss       last          counter      weight    label  predict features
> 4.000000   4.000000            1         1.0   2.0000   0.0000      104
> ...
> 1.118318   0.962480       524288    524288.0   2.0000   2.0007        5
> 
> finished run
> number of examples per pass = 807449
> passes used = 1
> weighted example sum = 807449
> weighted label sum = 1.6647e+06
> average loss = 1.01294
> best constant = 2.06168
> total feature number = 22483110

output models are exactly the same:

> diff out.readable out_l1.readable

However, with an older version from one of my colleagues, vw behaves as I would expect:

> old_vw --sgd  --noconstant -c -f out_l1 --l1 100 --passes 1  --readable_model out_l1.readable -k < x0
> using l1 regularization = 100
> final_regressor = out_l1
> Num weight bits = 18
> learning rate = 10
> initial_t = 1
> power_t = 0.5
> creating cache_file = .cache
> Reading datafile = 
> num sources = 1
> average    since         example     example  current  current  current
> loss       last          counter      weight    label  predict features
> 4.000000   4.000000            1         1.0   2.0000   0.0000      104
> ...
> 4.248032   4.324745       524288    524288.0   2.0000   0.0000        5
> 
> finished run
> number of examples per pass = 807449
> passes used = 1
> weighted example sum = 807449
> weighted label sum = 1.6647e+06
> average loss = 4.30842
> best constant = 2.06168
> total feature number = 22483110

Here's the version from the old binary

> old_vw --version
> 7.4.0
